Libplanet.MySQLStore/MySQLStore.cs

```diff
@@ -0,0 +1,806 @@
+        {
+            byte[] prefix = TxNonceKeyPrefix;
+
+            using QueryFactory db = MySqlUtils.OpenMySqlDB(_connectionString, _compiler);
+            Query query = db.Query(ChainDbName).Select().Where("Prefix", prefix);
+            IEnumerable<ChainModel> txNonces = query.Where("Cf", chainId.ToString()).Get<ChainModel>();
+
+            foreach (var txNonce in txNonces)
+            {
+                byte[] addressBytes = txNonce.Key.Skip(prefix.Length).ToArray();
+                Address address = new Address(addressBytes);
+                long nonce = MySqlStoreBitConverter.ToInt64(txNonce.Value);
+                yield return new KeyValuePair<Address, long>(address, nonce);
+            }
+        }
+
+        /// <inheritdoc/>
+        public override long GetTxNonce(Guid chainId, Address address)
+        {
+            byte[] key = TxNonceKey(address);
+            IList<ChainModel> chain = Select<ChainModel, byte[]>(ChainDbName, "Key", key);
+
+            byte[] bytes = null;
+
+            if (chain.Any())
+            {
+                foreach (var c in chain)
+                {
+                    if (c.Cf == chainId.ToString())
+                    {
+                        bytes = c.Value;
+                    }
+                }
+            }
+
+            return bytes is null
+                ? 0
+                : MySqlStoreBitConverter.ToInt64(bytes);
+        }
+
+        /// <inheritdoc/>
+        public override void IncreaseTxNonce(Guid chainId, Address signer, long delta = 1)
+        {
+            long nextNonce = GetTxNonce(chainId, signer) + delta;
+
+            byte[] key = TxNonceKey(signer);
+            byte[] bytes = MySqlStoreBitConverter.GetBytes(nextNonce);
+
+            QueryFactory db = new QueryFactory(new MySqlConnection(_connectionString), _compiler);
+
+            db.Statement(
+                "INSERT INTO Chain (`key`, `value`, `cf`, `prefix`) VALUES (@key, @value, @cf, @prefix) " +
+                "ON DUPLICATE KEY UPDATE `key` = VALUES(`key`), `value` = VALUES(`value`), `cf` = VALUES(`cf`), `prefix` = VALUES(`prefix`)",
+                new
+                {
+                    key = key,
+                    value = bytes,
+                    cf = chainId.ToString(),
+                    prefix = TxNonceKeyPrefix,
+                });
+        }
+
+        /// <inheritdoc/>
+        public override long CountTransactions()
+        {
+            return IterateTransactionIds().LongCount();
+        }
+
+        /// <inheritdoc/>
+        public override long CountBlocks()
+        {
+            return IterateBlockHashes().LongCount();
+        }
+
+        public override void Dispose()
+        {
+            QueryFactory db = new QueryFactory(new MySqlConnection(_connectionString), _compiler);
+
+            db.Statement($"SELECT Concat('TRUNCATE TABLE ', TABLE_NAME)" +
+                "FROM INFORMATION_SCHEMA.TABLES WHERE table_schema = 'libplanet.mysql'");
```

Comment:
Could you describe your intention from this query, please?

Reply:
My understanding of `Dispose()` was to clean all the data in the database so the query truncates all tables in `libplanet.mysql` database. Is that the right intention of `Dispose()`?

Reply:
In my understanding, `Dispose()` should be used for releasing files, handles or cleaning something.

As quotation from *.NET fundamentals*:

> There are additional reasons for implementing Dispose, for example, to free memory that was allocated, remove an item that was added to a collection, or signal the release of a lock that was acquired.

So, IMO what the `IDisposable.Dispose()` implementation should do is to dispose other variables like connections(i.e., `_db`).

Additionally, what I wondered at these lines was that it collects the results from `CONCAT` but they didn't use anywhere.

## References

https://docs.microsoft.com/en-us/dotnet/api/system.idisposable.dispose?view=net-5.0&viewFallbackFrom=net-3.1
https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose

Reply:
Thanks for your feedback! I'll make the correct implementation for `IDispose()`.

As for "_Additionally, what I wondered at these lines was that it collects the results from CONCAT but they didn't use anywhere_", I thought the code would do get all the tables and truncate them.
Something like: 
```
            List<TableModel> tables = Select<TableModel, string>("INFORMATION_SCHEMA.TABLES", "table_schema", "libplanet.mysql");
            foreach (var table in tables)
            {
                db.Statement($"TRUNCATE TABLE {table.table_name}");
            } 
```
But I realized only returns the table_name and doesn't truncate the tables 😅 .